contracts/staking/AbstractStaking.sol

```diff
@@ -0,0 +1,200 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.4;
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+import {AbstractValueDistributor} from "./AbstractValueDistributor.sol";
+
+/**
+ * @notice The AbstractStaking module
+ *
+ * Contract module for staking tokens and earning rewards based on shares.
+ */
+abstract contract AbstractStaking is AbstractValueDistributor, Initializable {
+    using SafeERC20 for IERC20;
+
+    address private _sharesToken;
+    address private _rewardsToken;
+
+    /**
+     * @dev The rate of rewards distribution per second.
+     *
+     * It determines the rate at which rewards are earned and distributed
+     * to stakers based on their shares.
+     *
+     * Note: Ensure that the `_rate` value is set correctly to match
+     * the decimal precision of the `_rewardsToken` to ensure accurate rewards distribution.
+     */
+    uint256 private _rate;
+
+    uint256 private _stakingStartTime;
+
+    /**
+     * @dev Throws if the staking has not started yet.
+     */
+    modifier stakingStarted() {
+        _checkStakingStarted();
+        _;
+    }
+
+    /**
+     * @notice Initializes the contract setting the values provided as shares token, rewards token, reward rate and staking start time.
+     * @param sharesToken_ The address of the shares token.
+     * @param rewardsToken_ The address of the rewards token.
+     * @param rate_ The reward rate.
+     * @param stakingStartTime_ The staking start time
+     */
+    function __AbstractStaking_init(
+        address sharesToken_,
+        address rewardsToken_,
+        uint256 rate_,
+        uint256 stakingStartTime_
+    ) internal onlyInitializing {
+        require(sharesToken_ != address(0), "Staking: zero address cannot be the Shares Token");
+        require(rewardsToken_ != address(0), "Staking: zero address cannot be the Rewards Token");
```

Comment:
When shares and reward tokens are equal, we may accidently withdraw other users shares as a reward. Please, leave warning if it is an expected behavior.

Reply:
Let's leave a comment on that. Virtually it is not a smart contract issue